rclcpp_lifecycle/src/transition.cpp

```diff
@@ -96,16 +108,110 @@ Transition::Transition(
   transition_handle_ = const_cast<rcl_lifecycle_transition_t *>(rcl_lifecycle_transition_handle);
 }
 
+Transition::Transition(const Transition & rhs)
+: allocator_(rhs.allocator_),
+  owns_rcl_transition_handle_(false),
+  transition_handle_(nullptr)
+{
+  *this = rhs;
+}
+
 Transition::~Transition()
 {
   reset();
 }
 
+Transition &
+Transition::operator=(const Transition & rhs)
+{
+  if (this == &rhs) {
+    return *this;
+  }
+
+  // reset all currently used resources
+  reset();
+
+  allocator_ = rhs.allocator_;
+  owns_rcl_transition_handle_ = rhs.owns_rcl_transition_handle_;
+
+  // we don't own the handle, so we can return straight ahead
+  if (!owns_rcl_transition_handle_) {
+    transition_handle_ = rhs.transition_handle_;
+    return *this;
+  }
+
+  // we own the handle, so we have to deep-copy the rhs object
+  transition_handle_ = static_cast<rcl_lifecycle_transition_t *>(
+    allocator_.allocate(sizeof(rcl_lifecycle_transition_t), allocator_.state));
```

Comment:
So I noticed that the allocation here's matching deallocation is in the fini function for this struct (same for above rcl struct in State class). While this isn't wrong, it's definitely different from how the rest of the fini functions work. Rather than iterate over the internet, I'm giving @Karsten1987 the ok to merge (since it's not broken, just weird) and tomorrow I'll help him clean up the implementation.